### PR TITLE
Category controller and required layers added

### PR DESCRIPTION
**Pull request by validatorr**

### CODE REVIEW

**Review by evgenkolner — APPROVED** [review]
good decision

File: src/main/java/v/hryhoryk/onlinebookstore/controller/CategoriesController.java [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    @Operation(summary = "Delete category by id",
+            description = "Admin can delete category by id")
+    @ApiResponses(value = {
+            @ApiResponse(responseCode = "200",
+                    description = "Successfully deleted"),
+            @ApiResponse(responseCode = "404",
+                    description = "Not found - category with this id is not existing")
+    })
+    @PreAuthorize("hasRole('ADMIN')")
+    @DeleteMapping("/{id}")
```
Comment (liliia-ponomarenko) [inline src/main/java/v/hryhoryk/onlinebookstore/controller/CategoriesController.java:111]: Let’s  add response status NO_CONTENT

File: src/main/java/v/hryhoryk/onlinebookstore/controller/CategoriesController.java [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+            @ApiResponse(responseCode = "200",
+                    description = "Successfully retrieved"),
+            @ApiResponse(responseCode = "404",
+                    description = "Not found - there are no categories available")
+    })
+    @PreAuthorize("hasRole('USER')")
+    @GetMapping("/{id}")
+    public CategoryResponseDto getCategoryById(
+            @PathVariable @Positive Long id) {
+        return categoryService.getById(id);
```
Comment (liliia-ponomarenko) [inline src/main/java/v/hryhoryk/onlinebookstore/controller/CategoriesController.java:83]: I can find that this endpoint is required but you can leave it;)

File: src/main/java/v/hryhoryk/onlinebookstore/dto/bookdto/CreateBookRequestDto.java [hunk 3]
```diff
@@ -21,5 +22,8 @@ public record CreateBookRequestDto(
         String description,
         @URL(message =
                 "It's required to add image of book to increase customer's trust rate")
-        String coverImage
-) {}
+        String coverImage,
+        @NotNull(message = "Make sure that book has it's category")
```
Comment (liliia-ponomarenko) [inline src/main/java/v/hryhoryk/onlinebookstore/dto/bookdto/CreateBookRequestDto.java]: Maybe NotEmpty

File: src/main/java/v/hryhoryk/onlinebookstore/repository/book/BookRepository.java [hunk 4]
```diff
@@ -1,10 +1,23 @@
...
 import org.springframework.data.jpa.repository.JpaRepository;
 import org.springframework.data.jpa.repository.JpaSpecificationExecutor;
+import org.springframework.data.jpa.repository.Query;
 import org.springframework.stereotype.Repository;
 import v.hryhoryk.onlinebookstore.model.Book;
 
 @Repository
 public interface BookRepository extends JpaRepository<Book, Long>, JpaSpecificationExecutor<Book> {
+    @EntityGraph(attributePaths = "categories")
+    Optional<Book> findById(Long id);
```
Comment (liliia-ponomarenko) [inline src/main/java/v/hryhoryk/onlinebookstore/repository/book/BookRepository.java]: Do you need a book with categories?

File: src/main/java/v/hryhoryk/onlinebookstore/repository/book/BookRepository.java [hunk 5]
```diff
@@ -1,10 +1,23 @@
...
+
+    @Query(value = "from Book b join fetch b.categories c where c.id = :id")
+    List<Book> findBooksByCategoryId(Long id);
+
+    @Query(value = "from Book b join fetch b.categories where b.isDeleted = false")
```
Comment (liliia-ponomarenko) [inline src/main/java/v/hryhoryk/onlinebookstore/repository/book/BookRepository.java:16]: the same how it works without query?

File: src/main/java/v/hryhoryk/onlinebookstore/service/book/BookServiceImpl.java [hunk 6]
```diff
@@ -56,10 +65,24 @@ public void deleteById(Long id) {
...
+    @Override
+    public List<BookDtoWithoutCategoryIds> getBooksByCategoryId(Long id) {
+        return bookRepository.findBooksByCategoryId(id)
+                .stream()
+                .map(bookMapper::toDtoWithoutCategories)
                 .toList();
     }
+
+    private void setCategories(Book book, List<Long> categoryIds) {
+        Set<Category> categories = new HashSet<>(categoryRepository.findAllById(categoryIds));
```
Comment (liliia-ponomarenko) [inline src/main/java/v/hryhoryk/onlinebookstore/service/book/BookServiceImpl.java]: you can use getReferenceById (HINT: use Transactional  for fixing LIE)

File: src/main/java/v/hryhoryk/onlinebookstore/service/category/CategoryServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+                .map(categoryMapper::toDto)
+                .toList();
+    }
+
+    @Override
+    public CategoryResponseDto getById(Long id) {
+        return categoryRepository.findById(id)
+                .map(categoryMapper::toDto)
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Can't find categoryIds with id: " + id));
```
Comment (liliia-ponomarenko) [inline src/main/java/v/hryhoryk/onlinebookstore/service/category/CategoryServiceImpl.java]:
```suggestion
                        "Can't find category with id: " + id));
```

File: src/main/java/v/hryhoryk/onlinebookstore/service/category/CategoryServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    @Override
+    public CategoryResponseDto createCategory(CreateCategoryRequestDto categoryDto) {
+        Category categoryEntity = categoryMapper.toCategory(categoryDto);
+        return categoryMapper.toDto(categoryRepository.save(categoryEntity));
+    }
+
+    @Override
+    public CategoryResponseDto updateById(Long id, CreateCategoryRequestDto categoryDto) {
+        Category categoryFromDb = categoryRepository.findById(id).orElseThrow(
+                () -> new EntityNotFoundException("Can't find categoryIds with id: " + id));
```
Comment (liliia-ponomarenko) [inline src/main/java/v/hryhoryk/onlinebookstore/service/category/CategoryServiceImpl.java]: the same

File: src/main/resources/db/changelog/changes/05-create-categories-table.yaml [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+                    nullable: false
+              - column:
+                  name: description
+                  type: varchar(255)
+              - column:
+                  name: is_deleted
+                  type: boolean
+                  defaultValueBoolean: false
+                  constraints:
+                    nullable: false
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/changes/05-create-categories-table.yaml]: add one empty line

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let’s  improve your solution ;)

File: src/main/java/v/hryhoryk/onlinebookstore/repository/book/BookRepository.java (hunk 5 above)
Comment (validatorr) [reply src/main/java/v/hryhoryk/onlinebookstore/repository/book/BookRepository.java:16]: Nope, without query it's not working at all

**Review by Elena-Bruyako — APPROVED** [review]

**Pull request closed without merging**